Tabbed forms built using customization framework

GPaoloni opened the pull request:
Jira: https://bugs.benetech.org/browse/CHI-439

This PR is grouping the following PRs: https://github.com/tech-matters/flex-plugins/pull/256, https://github.com/tech-matters/flex-plugins/pull/257, https://github.com/tech-matters/flex-plugins/pull/258, https://github.com/tech-matters/flex-plugins/pull/273, https://github.com/tech-matters/flex-plugins/pull/275, https://github.com/tech-matters/flex-plugins/pull/276, https://github.com/tech-matters/flex-plugins/pull/280, https://github.com/tech-matters/flex-plugins/pull/285, https://github.com/tech-matters/flex-plugins/pull/286, https://github.com/tech-matters/flex-plugins/pull/289, in order to introduce customization framework and use it in `TabbedForms` components as a single working feature.

nickhurlburt commented:
Something's weird with the alignment of categories on contact detail:
![image](https://user-images.githubusercontent.com/10714292/103247407-a5d3d500-491b-11eb-8f92-6f002640c34c.png)

This seems to affect both contacts added with these new forms, and older contacts.

nickhurlburt commented:
This is in Contact Details.  We shouldn't display 'mixed' to the user:
![image](https://user-images.githubusercontent.com/10714292/103247581-31e5fc80-491c-11eb-99f4-28929efa624c.png)

nickhurlburt commented:
When first accepting a task and loading the contact form, there's a console warning.  I don't know for sure if it is new as of this PR:
```
Warning: Received `true` for a non-boolean attribute `display`.

If you want to write it to the DOM, pass a string instead: display="true" or display={value.toString()}.
    in div (created by TabbedFormTabContainer)
    in TabbedFormTabContainer (at TabbedFormTab.tsx:48)
    in TabbedFormTab (created by Connect(TabbedFormTab))
    in Connect(TabbedFormTab) (created by Context.Consumer)
    in WithTaskContext(Connect(TabbedFormTab)) (at TabbedForms.tsx:158)
    in div (at TabbedForms.tsx:150)
    in div (created by TabbedFormsContainer)
    in TabbedFormsContainer (at TabbedForms.tsx:140)
    in div (at TabbedForms.tsx:139)
    in FormProvider (at TabbedForms.tsx:138)
    in TabbedForms (created by Connect(TabbedForms))
    in Connect(TabbedForms) (created by Context.Consumer)
    in WithTaskContext(Connect(TabbedForms)) (at HrmForm.tsx:30)
    in HrmForm (created by Connect(HrmForm))
    in Connect(HrmForm) (created by Context.Consumer)
    in WithTaskContext(Connect(HrmForm)) (at TaskView.jsx:63)
    in div (at TaskView.jsx:61)
    in TaskView (created by Connect(TaskView))
    in Connect(TaskView) (created by Context.Consumer)
    in WithTaskContext(Connect(TaskView)) (at CustomCRMContainer.tsx:35)
    in div (created by Absolute)
    in Absolute (at CustomCRMContainer.tsx:33)
    in CustomCRMContainer (created by Connect(CustomCRMContainer))
    in Connect(CustomCRMContainer) (at setUpComponents.js:250)
```

nickhurlburt commented:
You probably already know about this but this is currently storing any string as a phone number:
![image](https://user-images.githubusercontent.com/10714292/103251636-51862080-492e-11eb-9965-ea38be220c69.png)

When we merge this PR it's okay for some of these things to be still missing, but it is important that we provide a list of "known bugs" so that people aren't wasting their time on testing something we know is not working.

nickhurlburt commented:
Except for the issues above, I haven't found anything else.

andvirga commented on plugin-hrm-form/src/components/common/forms/categoriesTabGenerator.tsx:
```diff
@@ -0,0 +1,131 @@
+) => {
+  const path = [...parents, subcategory].join('.');
+
+  return (
+    <ConnectForm key={path}>
+      {({ register, getValues }) => {
+        const { categories } = getValues();
+        const checked = categories && categories.includes(path);
+        const disabled = categories && categories.length >= 3 && !checked;
+        const lighterColor = `${color}99`; // Hex with alpha 0.6
+
+        return (
+          <CategoryCheckboxLabel>
+            <CategoryCheckboxField color={lighterColor} selected={checked} disabled={disabled}>
+              <CategoryCheckbox
+                key={`${path}-checkbox`}
+                type="checkbox"
+                name="categories"
+                value={path}
+                onChange={updateCallback}
+                innerRef={register({ required: true, minLength: 1, maxLength: 3 })}
+                disabled={disabled}
+                color={lighterColor}
+              />
+              {subcategory}
+            </CategoryCheckboxField>
+          </CategoryCheckboxLabel>
+        );
+      }}
+    </ConnectForm>
+  );
+};
+
```
This is a minor UI bug that can be addressed in another ticket:
1. Open one "contactless" task
2. Open one chat task
3. Start filling the form in the two tasks
4. For chat task select three random categories (for example the first three subcategories from "Missing Children").
5. For contactless task select other three random categories (for example the first three subcategories from "Violence").
6. If you switch to chat task again, the three selected subcategories will appear with slightly different colors.

![image](https://user-images.githubusercontent.com/12242771/103285261-64473680-49bc-11eb-8abf-5316b8fc382b.png)

andvirga reviewed (approved):
I haven't reviewed the code, since we already did it in the multiple PRs that you've opened. 
However, I've tested it and only found one minor issue. This is excellent work!

GPaoloni commented:
@nickhurlburt I've fixed both errors you spotted, on being a simple style change, the other one a mistake when transforming the form before sending it to the backend. The proptype warning was there I believe, I've seen that many times, and in fact I've removed some of them, not introduced any new ones.
@andvirga I'll address that UI bug later on, as a separate task, so I can wrap this one asap.
Thanks to both!!
